src/hooks/useImmediateState.ts

```diff
@@ -0,0 +1,32 @@
+import { useRef, useEffect, useState, useCallback } from 'react';
+
+type SetStateAction<S> = S | ((prevState: S) => S);
+type SetState<S> = (action: SetStateAction<S>) => void;
+
+export const useImmediateState = <S extends {}>(init: S): [S, SetState<S>] => {
+  const isMounted = useRef(false);
+  const initialState = useRef<S>({ ...init });
+  const [state, setState] = useState<S>(initialState.current);
+
+  const updateState: SetState<S> = useCallback((action: SetStateAction<S>) => {
```

Comment:
It looks like this does the following:
 - on first render
 - hijack `setState` call
 - uses mutation to update state value

Is that the case?

Reply:
Ehhh, yes if you mean wrapping it by hijacking 👍😅

Reply:
Updated!

I found this pretty tricky to understand but I see what you're doing now. I think the same could be achieved by currying the executeQuery function and passing in:
 - A mutation setState function on initial render
 - The actual setState function on further renders

That would get rid of the need for an abstraction.

```tsx
const useQuery = () => {
    const isMounted = useRef(false);
    const initialState = useRef({ ... });
    const [state, setState] = useState(initialState);

    const updateState = useMemo(() => isMounted.current ? setState : (arg) => Object.assign(initialState, arg), [isMounted.current]);

    const executeQuery = useMemo(() => doExecuteQuery(updateState), [updateState])
    
    // ...
}

const doExecuteQuery = (setState, otherArgs) => () => // ...
```

Reply:
It's not quite the same as you'd then trigger an update since `executeQuery` changes, since `updateState` changes after `initialMount` changes... yea, long chain of stuff happening. This is also not a deterministic update since it's depending on a ref, which means we get an unexpected query on the next update, whenever it comes 😅 anyway, since this is not about the code snippet:

I wouldn't inline this as the abstraction is nice imho; I think it reduces the amount of complexity for external contributors since they're just looking at drop-in replacements for `useEffect` and `useState` that they can assume to work the same.

Given that all bugs / things we've now fixed are really subtle variations on how one would usually write these features or this behaviour, the hooks ensure that we can continue to write them normally, mostly/hopefully bug-free, without making our code take on more concerns.